lib/CodeTransform/Adapter/WorseReflection/Refactor/WorsePromoteProperty.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Phpactor\CodeTransform\Adapter\WorseReflection\Refactor;
+
+use Microsoft\PhpParser\Node\MethodDeclaration;
+use Microsoft\PhpParser\Node\Parameter;
+use Microsoft\PhpParser\Parser;
+use Phpactor\CodeBuilder\Domain\Builder\SourceCodeBuilder;
+use Phpactor\CodeBuilder\Domain\Code;
+use Phpactor\CodeBuilder\Domain\Prototype\Visibility;
+use Phpactor\CodeBuilder\Domain\Updater;
+use Phpactor\CodeTransform\Domain\Refactor\PromoteProperty;
+use Phpactor\TextDocument\ByteOffset;
+use Phpactor\TextDocument\TextDocument;
+use Phpactor\TextDocument\TextEdits;
+use Phpactor\WorseReflection\Bridge\TolerantParser\Reflection\ReflectionMethod;
+use Phpactor\WorseReflection\Core\Exception\NotFound;
+use Phpactor\WorseReflection\Core\SourceCode;
+use Phpactor\WorseReflection\Reflector;
+
+class WorsePromoteProperty implements PromoteProperty
+{
+    public function __construct(
+        private Reflector $reflector,
+        private Parser $parser,
+        private Updater $updater,
+    ) {
+    }
+
+    public function promoteProperty(TextDocument $document, ByteOffset $offset): TextEdits
+    {
+        $node = $this->parser->parseSourceFile($document->__toString())->getDescendantNodeAtPosition($offset->toInt());
+        $node = $node->getFirstAncestor(MethodDeclaration::class);
+
+        if (!$node instanceof MethodDeclaration) {
+            return TextEdits::none();
+        }
+
+        if ($node->getName() !== '__construct') {
+            return TextEdits::none();
+        }
+
+        if(count($node->parameters->children) === 0) {
+            return TextEdits::none();
+        }
+
+        $constructor = $this->reflector->reflectNode($document, $node->getStartPosition());
+        if (!$constructor instanceof ReflectionMethod) {
+            return TextEdits::none();
+        }
+
+        // Todo stuff
+    }
```

Comment:
I need help with this function I am not sure what the best way is. Ideas that I have:

Get the `Parameter` node and then use the `getFirstAncestor` again to get the `MethodDeclaration` for name checking. And then I can still use the original node I got for all the manipulation. The problem is that we need the class name to let the source code updater know which class to update the constructor on. Any ideas?

Reply:
Can you not copy the existing transformer for complete constructor? 

https://github.com/phpactor/phpactor/blob/474eb3ccd7c58af5b7f83466dadaa88ce0eec6e6/lib/CodeTransform/Adapter/WorseReflection/Transformer/CompleteConstructor.php#L36-L40

This is very similar

Reply:
I just had a look at this. The thing is I would recommend to change the `CompleteConstructor`. It would make the command easier, generates less code and would also do the job.

The thing that I found with the `PromoteProperty` action is that if you already have a non promoted property with the property separately you also have to modify the method's body to remove the assignment of the property, which the `SourcecodeBuilder` doesn't support yet.

Reply:
In care it has an assignment we should automatically leave it alone, less WTF :)

Reply:
I mean you could make the case of both, because leaving it is just asking for trouble if the promoted argument is getting renamed.

Reply:
not sure I understand: I think for the sake of simplicity, this action should only act on parameters without visibility which have no corresponding property, as (for me at least) that is 99% of the use case, There are things like PHP-CS-Fixer and Rector which can migrate assigned properties to promoted ones.

Reply:
Yeah, as you figured out below. Technically if that's the only use case you could just add a shortcut to vim like `Iprivate ` or `Ipublic ` if you are already on the parameter. :D But having it in phpactor would be much nicer, I agree.